filters/auth/oidc.go

```diff
@@ -0,0 +1,561 @@
+	)
+
+	// filter specific checks
+	switch f.typ {
+	case checkOIDCUserInfo:
+		var container userInfoContainer
+		err := json.Unmarshal([]byte(cookie), &container)
+		if err != nil {
+			unauthorized(ctx, "failed to deserialize cookie: "+err.Error(), invalidToken, r.Host)
+			return
+		}
+		if container.OAuth2Token.Valid() && container.UserInfo != nil {
+			allowed = true
+		}
+		sub = container.Subject
+	case checkOIDCAnyClaims:
+		var container claimsContainer
+		err := json.Unmarshal([]byte(cookie), &container)
+		if err != nil {
+			unauthorized(ctx, "failed to deserialize cookie: "+err.Error(), invalidToken, r.Host)
+			return
+		}
+		allowed = f.validateAnyClaims(container.Claims)
+		log.Infof("validateAnyClaims: %v", allowed)
+		sub = container.Subject
+	case checkOIDCAllClaims:
+		var container claimsContainer
+		err := json.Unmarshal([]byte(cookie), &container)
+		if err != nil {
+			unauthorized(ctx, "failed to deserialize cookie: "+err.Error(), invalidToken, r.Host)
+			return
+		}
+		allowed = f.validateAllClaims(container.Claims)
+		log.Infof("validateAllClaims: %v", allowed)
+		sub = container.Subject
+	default:
+		unauthorized(ctx, "unknown", invalidFilter, r.Host)
+		return
+	}
+
+	if !allowed {
+		log.Infof("unauthorized")
+		unauthorized(ctx, "insufficient permissions :"+sub, invalidClaim, r.Host)
+		return
+	}
+
+	log.Infof("send authorized")
+	authorized(ctx, sub)
+}
+
+func (f *tokenOidcFilter) tokenClaims(ctx filters.FilterContext, oauth2Token *oauth2.Token) (map[string]interface{}, string, error) {
+	r := ctx.Request()
+	rawIDToken, ok := oauth2Token.Extra("id_token").(string)
+	if !ok {
+		return nil, "", fmt.Errorf("invalid token, no id_token field in oauth2 token")
+	}
+
+	var idToken *oidc.IDToken
+	idToken, err := f.verifier.Verify(r.Context(), rawIDToken)
+	if err != nil {
+		return nil, "", err
+	}
+
+	tokenMap := make(map[string]interface{})
+	if err = idToken.Claims(&tokenMap); err != nil {
+		return nil, "", err
+	}
+
+	sub, ok := tokenMap["sub"].(string)
+	if !ok {
+		return nil, "", fmt.Errorf("claims does not contain sub")
+	}
+	return tokenMap, sub, nil
+}
+
+func (f *tokenOidcFilter) getCallbackState(ctx filters.FilterContext) (*OauthState, error) {
+	// CSRF protection using similar to
+	// https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)_Prevention_Cheat_Sheet#Encrypted_Token_Pattern,
+	// because of https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest
+	r := ctx.Request()
+	stateQueryEncHex := r.URL.Query().Get("state")
+	if stateQueryEncHex == "" {
+		return nil, fmt.Errorf("no state parameter")
+	}
+
+	stateQueryEnc := make([]byte, len(stateQueryEncHex))
+	if _, err := fmt.Sscanf(stateQueryEncHex, "%x", &stateQueryEnc); err != nil && err != io.EOF {
+		log.Errorf("Failed to read hex string: %v", err)
+		return nil, err
+	}
+
+	stateQueryPlain, err := f.encrypter.decryptDataBlock(stateQueryEnc)
+	if err != nil {
+		log.Errorf("token from state query is invalid: %v", err)
```

Comment:
This log can be triggered by the client if I understand correctly.
The log is ok but please comment that we maybe want to have a metric for this.

Reply:
Added a `TODO`